File: Visual Studio Product Team/ADO.NET Entity Framework/[C#]-ADO.NET Entity Framework/C#/EmployeeTracker/ViewModel/DepartmentWorkspaceViewModel.cs
// Copyright © Microsoft Corporation.  All Rights Reserved.
// This code released under the terms of the 
// Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)


namespace EmployeeTracker.ViewModel
{
    using System;
    using System.Collections.ObjectModel;
    using System.Windows.Input;
    using EmployeeTracker.Common;
    using EmployeeTracker.Model;
    using EmployeeTracker.ViewModel.Helpers;

    /// <summary>
    /// ViewModel for managing Departments within the company
    /// </summary>
    public class DepartmentWorkspaceViewModel : ViewModelBase
    {
        /// <summary>
        /// The deprtment currently selected in the workspace
        /// </summary>
        private DepartmentViewModel currentDepartment;

        /// <summary>
        /// UnitOfWork for managing changes
        /// </summary>
        private IUnitOfWork unitOfWork;

        /// <summary>
        /// Initializes a new instance of the DepartmentWorkspaceViewModel class.
        /// </summary>
        /// <param name="departments">The departments to be managed</param>
        /// <param name="unitOfWork">UnitOfWork for managing changes</param>
        public DepartmentWorkspaceViewModel(ObservableCollection<DepartmentViewModel> departments, IUnitOfWork unitOfWork)
        {
            if (departments == null)
            {
                throw new ArgumentNullException("departments");
            }

            if (unitOfWork == null)
            {
                throw new ArgumentNullException("unitOfWork");
            }

            this.unitOfWork = unitOfWork;
            this.AllDepartments = departments;
            this.CurrentDepartment = this.AllDepartments.Count > 0 ? this.AllDepartments[0] : null;

            // Re-act to any changes from outside this ViewModel
            this.AllDepartments.CollectionChanged += (sender, e) =>
            {
                if (e.OldItems != null && e.OldItems.Contains(this.CurrentDepartment))
                {
                    this.CurrentDepartment = null;
                }
            };

            this.AddDepartmentCommand = new DelegateCommand((o) => this.AddDepartment());
            this.DeleteDepartmentCommand = new DelegateCommand((o) => this.DeleteCurrentDepartment(), (o) => this.CurrentDepartment != null);
        }

        /// <summary>
        /// Gets the command for adding a new department
        /// </summary>
        public ICommand AddDepartmentCommand { get; private set; }

        /// <summary>
        /// Gets the command for deleting the current department
        /// </summary>
        public ICommand DeleteDepartmentCommand { get; private set; }

        /// <summary>
        /// Gets all departments whithin the company
        /// </summary>
        public ObservableCollection<DepartmentViewModel> AllDepartments { get; private set; }

        /// <summary>
        /// Gets or sets the deprtment currently selected in the workspace
        /// </summary>
        public DepartmentViewModel CurrentDepartment
        {
            get
            {
                return this.currentDepartment;
            }

            set
            {
                this.currentDepartment = value;
                this.OnPropertyChanged("CurrentDepartment");
            }
        }

        /// <summary>
        /// Handles addition a new department to the workspace and model
        /// </summary>
        private void AddDepartment()
        {
            Department dep = this.unitOfWork.CreateObject<Department>();
            this.unitOfWork.AddDepartment(dep);

            DepartmentViewModel vm = new DepartmentViewModel(dep);
            this.AllDepartments.Add(vm);
            this.CurrentDepartment = vm;
        }

        /// <summary>
        /// Handles deletion of the current department
        /// </summary>
        private void DeleteCurrentDepartment()
        {
            this.unitOfWork.RemoveDepartment(this.CurrentDepartment.Model);
            this.AllDepartments.Remove(this.CurrentDepartment);
            this.CurrentDepartment = null;
        }
    }
}

File: Visual Studio Product Team/ADO.NET Entity Framework/[VB]-ADO.NET Entity Framework/VB/EmployeeTracker/ViewModel/DepartmentWorkspaceViewModel.vb
' Copyright © Microsoft Corporation.  All Rights Reserved.
' This code released under the terms of the 
' Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)



Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.ObjectModel
Imports System.Windows.Input
Imports EmployeeTracker.Common
Imports EmployeeTracker.Model
Imports EmployeeTracker.ViewModel.Helpers

Namespace EmployeeTracker.ViewModel

    ''' <summary>
    ''' ViewModel for managing Departments within the company
    ''' </summary>
    Public Class DepartmentWorkspaceViewModel
        Inherits ViewModelBase
        ''' <summary>
        ''' The deprtment currently selected in the workspace
        ''' </summary>
        Private privateCurrentDepartment As DepartmentViewModel

        ''' <summary>
        ''' UnitOfWork for managing changes
        ''' </summary>
        Private unitOfWork As IUnitOfWork

        ''' <summary>
        ''' Initializes a new instance of the DepartmentWorkspaceViewModel class.
        ''' </summary>
        ''' <param name="departments">The departments to be managed</param>
        ''' <param name="unitOfWork">UnitOfWork for managing changes</param>
        Public Sub New(ByVal departments As ObservableCollection(Of DepartmentViewModel), ByVal unitOfWork As IUnitOfWork)
            If departments Is Nothing Then
                Throw New ArgumentNullException("departments")
            End If

            If unitOfWork Is Nothing Then
                Throw New ArgumentNullException("unitOfWork")
            End If

            Me.unitOfWork = unitOfWork
            Me.AllDepartments = departments
            Me.CurrentDepartment = If(Me.AllDepartments.Count > 0, Me.AllDepartments(0), Nothing)

            ' Re-act to any changes from outside this ViewModel
            AddHandler Me.AllDepartments.CollectionChanged, Sub(sender, e)
                                                                If e.OldItems IsNot Nothing AndAlso e.OldItems.Contains(Me.CurrentDepartment) Then
                                                                    Me.CurrentDepartment = Nothing
                                                                End If
                                                            End Sub

            Me.AddDepartmentCommand = New DelegateCommand(Sub(o) Me.AddDepartment())
            Me.DeleteDepartmentCommand = New DelegateCommand(Sub(o) Me.DeleteCurrentDepartment(), Function(o) Me.CurrentDepartment IsNot Nothing)
        End Sub

        ''' <summary>
        ''' Gets the command for adding a new department
        ''' </summary>
        Private privateAddDepartmentCommand As ICommand
        Public Property AddDepartmentCommand() As ICommand
            Get
                Return privateAddDepartmentCommand
            End Get
            Private Set(ByVal value As ICommand)
                privateAddDepartmentCommand = value
            End Set
        End Property

        ''' <summary>
        ''' Gets the command for deleting the current department
        ''' </summary>
        Private privateDeleteDepartmentCommand As ICommand
        Public Property DeleteDepartmentCommand() As ICommand
            Get
                Return privateDeleteDepartmentCommand
            End Get
            Private Set(ByVal value As ICommand)
                privateDeleteDepartmentCommand = value
            End Set
        End Property

        ''' <summary>
        ''' Gets all departments whithin the company
        ''' </summary>
        Private privateAllDepartments As ObservableCollection(Of DepartmentViewModel)
        Public Property AllDepartments() As ObservableCollection(Of DepartmentViewModel)
            Get
                Return privateAllDepartments
            End Get
            Private Set(ByVal value As ObservableCollection(Of DepartmentViewModel))
                privateAllDepartments = value
            End Set
        End Property

        ''' <summary>
        ''' Gets or sets the deprtment currently selected in the workspace
        ''' </summary>
        Public Property CurrentDepartment() As DepartmentViewModel
            Get
                Return Me.privateCurrentDepartment
            End Get

            Set(ByVal value As DepartmentViewModel)
                Me.privateCurrentDepartment = value
                Me.OnPropertyChanged("CurrentDepartment")
            End Set
        End Property

        ''' <summary>
        ''' Handles addition a new department to the workspace and model
        ''' </summary>
        Private Sub AddDepartment()
            Dim dep As Department = Me.unitOfWork.CreateObject(Of Department)()
            Me.unitOfWork.AddDepartment(dep)

            Dim vm As New DepartmentViewModel(dep)
            Me.AllDepartments.Add(vm)
            Me.CurrentDepartment = vm
        End Sub

        ''' <summary>
        ''' Handles deletion of the current department
        ''' </summary>
        Private Sub DeleteCurrentDepartment()
            Me.unitOfWork.RemoveDepartment(Me.CurrentDepartment.Model)
            Me.AllDepartments.Remove(Me.CurrentDepartment)
            Me.CurrentDepartment = Nothing
        End Sub
    End Class
End Namespace
